### ISSUE TITLE
[discussion] Should Optimizers be also Modules?

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#58997]
Then their cached gradient moments can be their own "buffers" or "parameters", and then they may be used more conveniently in bi-level optimization models. Method `step()` would then call `forward()`, and also be callable from `__call__`

**Comment by SsnL** [comment]
In a sense they are metamodules- they take in (module, loss) and give you (updated_module,). It's just happens that the update is inplace.

**Comment by vadimkantorov** [comment]
E.g. solving inverse problems within the loop often needs some sort of optimizer. It would be cool to reuse them as module objects. It's of course already not bad that we have functional versions, but with sub-module API it may become even better

**Comment by vadimkantorov** [comment]
If modules could be hacked to work without necessary `super().__init__()` call (e.g. by introducing helper methods for accessing hooks to guard against non-existing attributes), then maybe optim.Optimizer could be simply derived from nn.Module :)

**Comment by vadimkantorov** [comment]
Probably all attributes could be made lazy properties (hooks, modules, parameters buffers, and that seems it), then the mandatory base constructor call could go? @ezyang

**Comment by ezyang** [comment]
Someone should hack up a prototype and see how it goes... hopefully it's not too much code to see if inheriting would work or not.

**Comment by albanD** [comment]
Thanks for the proposal!
I do agree that this is an interesting idea in the context of meta learning or inverse problems.

But I am not convinced that this is something we want for optimizers in general.
There are a few reasons for that:
- I think it can be confusing to users to have Optimizers be nn.Modules in the general case. You should definitely not try to make these part of your network main Module!
- It is not clear what should be a Parameter or Buffer there. Both when used in classical use cases, where nothing should be a Parameter? And in the meta learning case where what is a Parameter vs a Buffer would be problem specific.
- It is easy to make such a Module if you need it for your specific use case by using the functional API. And that's actually one of the reason for the function API to exist: if you need to handle the optimizer state differently, you can easily do it and still have access to the "step" function.

**Comment by vadimkantorov** [comment]
- I think it's good to try to remove the requirement of calling the base constructor regardless of optimizers <-> modules, especially if lazy properties allow to do that elegantly
- A question is for models where people already make optimizer a field of a module. Maybe worth coming up with some options for buffers / parameters that would keep them "private"? Also maybe make Parameter / Buffer derive from a common class? Or make a single class with some attributes set? Then the user may decide post-hoc if they want parameter or buffer behavior, private or not. Then there should be no confusion and any noticeable changes for existing code / users
- I think distinction between buffer / parameter not very useful, since anyway their updates are overridden by the optimizer (usually some moving average, side note - it would be good to have some "official" example of moving average of model params wrt no_grads/use of lerp etc - happens a lot now in byol-type models)

**Comment by vadimkantorov** [comment]
maybe also functional optimizers should have `inplace = False` mode (for allowing unrolling the updates if needed for bi-level optimization)
